clang/lib/Sema/SemaChecking.cpp

```diff
@@ -13809,21 +13809,30 @@ static void CheckImplicitConversion(Sema &S, Expr *E, QualType T,
 
       int Order = S.getASTContext().getFloatingTypeSemanticOrder(
           QualType(SourceBT, 0), QualType(TargetBT, 0));
+
       if (Order > 0) {
         // Don't warn about float constants that are precisely
         // representable in the target type.
         Expr::EvalResult result;
         if (E->EvaluateAsRValue(result, S.Context)) {
           // Value might be a float, a float vector, or a float complex.
-          if (IsSameFloatAfterCast(result.Val,
-                   S.Context.getFloatTypeSemantics(QualType(TargetBT, 0)),
-                   S.Context.getFloatTypeSemantics(QualType(SourceBT, 0))))
-            return;
+          if (IsSameFloatAfterCast(
+                  result.Val,
+                  S.Context.getFloatTypeSemantics(QualType(TargetBT, 0)),
+                  S.Context.getFloatTypeSemantics(QualType(SourceBT, 0)))) {
+            if (S.getLangOpts().SYCLIsDevice) {
+              S.SYCLDiagIfDeviceCode(CC, diag::warn_imp_float_size_conversion);
+              return;
+            } else {
+              DiagnoseImpCast(S, E, T, CC,
+                              diag::warn_imp_float_size_conversion);
+              return;
+            }
+          }
```

Comment:
```suggestion
          if (IsSameFloatAfterCast(
                  result.Val,
                  S.Context.getFloatTypeSemantics(QualType(TargetBT, 0)),
                  S.Context.getFloatTypeSemantics(QualType(SourceBT, 0)))) {
            if (S.getLangOpts().SYCLIsDevice)
              S.SYCLDiagIfDeviceCode(CC, diag::warn_imp_float_size_conversion);
            else
              DiagnoseImpCast(S, E, T, CC,
                              diag::warn_imp_float_size_conversion);
            
            return;
          }
```
Hoists the return out of both branches and drops some unnecessary curly braces; NFC.